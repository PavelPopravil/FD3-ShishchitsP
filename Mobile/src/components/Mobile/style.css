body {
  font-family: "Trebuchet MS", Arial, Helvetica, sans-serif;
  font-size:  16px;
}

* {
  box-sizing: border-box;
}

img {
  max-width: 100%;
  height: auto;
  display: block;
}

.container {
  max-width: 960px;
  margin: 0 auto;
  padding: 0 15px;
}

.mobile__clients tr + .mobile__clients tr {
  border-top: 1px solid #e5e5e5;
}

.mobile__clients {
  background-color: #fff;
  box-shadow: 0 8px 19.2px 0 rgba(0,0,0,.13);
  margin-bottom:  30px;
  width: 100%;
}

.mobile__clients thead {
  background-color: #d2e6f3;
}

.mobile__clients th,
.mobile__clients td {
  padding: 15px;
  text-align: center;
}

.btn {
  padding: 10px 15px;
  display: inline-block;
  background-color: #005fa9;
  border: 0;
  color: #fff;
  font-size:  16px;
  font-family: "Trebuchet MS", Arial, Helvetica, sans-serif;
  cursor: pointer;
  min-width: 100px;
  transition: all .5s ease;
}

.btn.disabled {
  background-color: #e5e5e5;
  cursor: not-allowed;
  pointer-events: none;
}

.delete-btn:hover {
  background-color: #005190;
}

.edit-btn {
  background-color: #ebac40;
}

.edit-btn:hover {
  background-color: #eb9b22;
}