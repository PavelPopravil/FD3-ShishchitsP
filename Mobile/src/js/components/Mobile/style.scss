.mobile {

  &__name {
    font-size: 40px;
  }

  &__clients {
    background-color: #fff;
    box-shadow: 0 8px 19.2px 0 rgba(0,0,0,.13);
    margin-bottom:  30px;
    width: 100%;
    font-size: 16px;
    border-collapse: collapse;
    border-spacing: 0;
    border: 1px solid #e5e5e5;

    thead {
      background-color: #d2e6f3;
    }

    tr + tr {
      border-top: 1px solid #e5e5e5;
    }

    th,
    td {
      padding: 15px;
      text-align: center;

      &.unavail {
        background-color: rgba(255, 0, 0, .4);
      }

      &.avail {
        background-color: rgba(0, 255, 0, .4);
      }

      a {
        text-decoration: none;
        color: #005190;

        &:hover {
          text-decoration: underline;
        }
      }
    }
  }
}

.btn {
  padding: 10px 15px;
  display: inline-block;
  background-color: #005fa9;
  border: 0;
  color: #fff;
  font-size:  16px;
  font-family: "Trebuchet MS", Arial, Helvetica, sans-serif;
  cursor: pointer;
  min-width: 100px;
  transition: all .5s ease;
}

.btn.disabled {
  background-color: #e5e5e5;
  cursor: not-allowed;
  pointer-events: none;
}

.delete-btn:hover {
  background-color: #005190;
}

.edit-btn {
  background-color: #ebac40;
}

.edit-btn:hover {
  background-color: #eb9b22;
}