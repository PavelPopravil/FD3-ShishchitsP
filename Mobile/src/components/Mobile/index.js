import React from 'react';
import PropTypes from 'prop-types';
import './style.scss';
import {appEvents} from '../event';

import MobileClient from '../MobileClient/index';
import MobileForm from '../MobileForm/index';

import * as mModules from '../../modules/mobile';

class Mobile extends React.PureComponent {

  static propTypes = {
    currCompanyName: PropTypes.string.isRequired,
    clients: PropTypes.arrayOf(
      PropTypes.shape({
        id: PropTypes.number.isRequired,
        name: PropTypes.string.isRequired,
        surName: PropTypes.string.isRequired,
        secondName: PropTypes.string.isRequired,
        balance: PropTypes.number.isRequired,
        status: PropTypes.number.isRequired // 0 - неактивен, 1 - активен
      }).isRequired
    ).isRequired
  };

  state = {
    clients: this.props.clients,
    filteredClients: this.props.clients,
    formMode: 0, // 1 - форма редактирования, 2 - форма добавления,
    clientToEdit: null,
    currentCompany: this.props.currCompanyName,
    filterMod: 'all'
  };

  componentDidMount() {
    appEvents.addListener('EItemRemove', this.removeClient);
    appEvents.addListener('EItemEdit', this.onClientEdit);
    appEvents.addListener('EEditClient', this.editClient);
    appEvents.addListener('EAddClient', this.addClient);
    appEvents.addListener('ECloseForm', this.closeForm);
  }

  componentWillUnmount() {
    appEvents.removeListener('EitemRemove', this.removeClient);
    appEvents.removeListener('EItemEdit', this.onClientEdit);
    appEvents.removeListener('EEditClient', this.editClient);
    appEvents.removeListener('EAddClient', this.addClient);
    appEvents.removeListener('ECloseForm', this.closeForm);
  }

  editClient = (editedClient) => {
    const editedClients = mModules.editClient(this.state.clients, editedClient);
    this.setState({
      clients: editedClients,
      filteredClients: this.filterList(editedClients),
      formMode: 0,
      clientToEdit: null
    });
  };

  addClient = (client) => {
    const addedClients = mModules.addClient(this.state.clients, client);
    this.setState({
      clients: addedClients,
      filteredClients: this.filterList(addedClients),
      formMode: 0,
      clientToEdit: null
    });
  };

  removeClient = (id) => {
    const clients = mModules.removeClient(this.state.clients, id);
    this.setState({clients: clients, filteredClients: this.filterList(clients), formMode: 0});
  };

  onClientEdit = (id) => {
    this.setState({formMode: 1, clientToEdit: this.state.clients.find((client) => client.id === id)});
  };

  onClientAdd = () => {
    const id = this.state.clients.length ? this.state.clients[this.state.clients.length - 1].id + 1 : 1;
    this.setState({formMode: 2, clientToEdit: {id: id, name: '', surName: '', secondName: '', balance: 0, status: ''}})
  };

  closeForm = (formMode) => {
    this.setState({formMode: formMode});
  };

  changeCompanyName = (e) => {
    this.setState({currentCompany: e.target.value})
  };

  filterList = (arr, type = this.state.filterMod) => {
    if (type === 'all') {
      return arr;
    } else {
      return arr.filter((item) => type === 'active' ? item.status === 1 : item.status === 0);
    }
  };

  onFilterClick = (e) => {
    const type = e.target.value;
    this.setState({filterMod: type, filteredClients: this.filterList(this.state.clients, type)})
  };

  render() {

    // console.log('MobileCompany render');

    const clients = this.state.filteredClients.map((client) => {
      return <MobileClient key={client.id} client={client}/>
    });

    return (
      <div className='mobile container'>
        <div className="mobile__name">{this.state.currentCompany}</div>
        <div className='mobile__btn-wrap'>
          <button className='btn' value='Velcom' onClick={this.changeCompanyName}>Velcom</button>
          <button className='btn' value='Mts' onClick={this.changeCompanyName}>Mts</button>
        </div>
        <div className='mobile__filter'>
          <button className='btn' value='all' onClick={this.onFilterClick}>Все</button>
          <button className='btn' value='active' onClick={this.onFilterClick}>Активные</button>
          <button className='btn' value='unavail' onClick={this.onFilterClick}>Заблокированные</button>
        </div>
        <table className='mobile__clients'>
          <thead>
          <tr>
            <td>Фамилия</td>
            <td>Имя</td>
            <td>Отчетсво</td>
            <td>Баланс</td>
            <td>Статус</td>
            <td>Управление</td>
          </tr>
          </thead>
          <tbody>
          {
            this.state.filteredClients.length ?
              clients
              :
              <tr>
                <td>Клиентов нет</td>
              </tr>
          }
          </tbody>
        </table>
        <div className='mobile__add'>
          <button className='btn edit-btn add-client' onClick={this.onClientAdd}>Добавить клиента</button>
        </div>
        <div className="mobile__form">
          {
            (this.state.formMode === 1) &&
            <MobileForm key={this.state.clientToEdit.id} btnText='Редактировать' title='Редактировать данные клиента'
                        client={this.state.clientToEdit} formMode={this.state.formMode}/>
          }
          {
            (this.state.formMode === 2) &&
            <MobileForm btnText='Добавить' title='Добавить нового клиента' client={this.state.clientToEdit}
                        formMode={this.state.formMode}/>
          }
        </div>
      </div>
    );
  }
}

export default Mobile;