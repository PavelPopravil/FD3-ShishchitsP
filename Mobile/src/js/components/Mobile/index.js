import React from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import './style.scss';
import isoFetch from 'isomorphic-fetch';
import { appEvents } from '../event';

import MobileClient from '../MobileClient/index';
import MobileForm from '../MobileForm/index';

import * as mModules from '../../modules/mobile';

import {fetchDataSuccess, fetchDataFailure} from '../../redux/actions/clientActions';

class Mobile extends React.PureComponent {

  static propTypes = {
    // path: PropTypes.string.isRequired,
    // clients: PropTypes.arrayOf(
    //   PropTypes.shape({
    //     id: PropTypes.number.isRequired,
    //     name: PropTypes.string.isRequired,
    //     surName: PropTypes.string.isRequired,
    //     secondName: PropTypes.string.isRequired,
    //     balance: PropTypes.number.isRequired,
    //     status: PropTypes.number.isRequired // 0 - неактивен, 1 - активен
    //   }).isRequired
    // ).isRequired
  };

  state = {
    clients: [],
    filteredClients: [],
    isLoaded: false,
    formMode: 0, // 1 - форма редактирования, 2 - форма добавления,
    clientToEdit: null,
    filterMod: 'all'
  };

  componentDidMount() {
    this.downloadData();
    // appEvents.addListener('EItemRemove', this.removeClient);
    // appEvents.addListener('EItemEdit', this.onClientEdit);
    // appEvents.addListener('EEditClient', this.editClient);
    // appEvents.addListener('EAddClient', this.addClient);
    // appEvents.addListener('ECloseForm', this.closeForm);
  }

  componentWillUnmount() {
    // appEvents.removeListener('EitemRemove', this.removeClient);
    // appEvents.removeListener('EItemEdit', this.onClientEdit);
    // appEvents.removeListener('EEditClient', this.editClient);
    // appEvents.removeListener('EAddClient', this.addClient);
    // appEvents.removeListener('ECloseForm', this.closeForm);
  }

  componentWillReceiveProps = (nextProps, nextContext) => {
    this.setState({clients: nextProps.clients, isLoaded: nextProps.isLoaded});
  };

  fetchError = (error) => {
    this.props.dispatch(fetchDataFailure(error));
  };

  fetchSuccess = (data) => {
    this.props.dispatch(fetchDataSuccess(data));
  };

  downloadData = () => {
    isoFetch("http://5c4ea42cd87cab001476ef73.mockapi.io/api/mts", {
      method: 'get',
      headers: {
        "Accept": "application/json",
      },
    })
      .then((response) => { // response - HTTP-ответ
        if (!response.ok) {
          let Err = new Error("fetch error " + response.status);
          Err.userMessage = "Ошибка связи";
          throw Err; // дальше по цепочке пойдёт отвергнутый промис
        } else
          return response.json(); // дальше по цепочке пойдёт промис с пришедшими по сети данными
      })
      .then((data) => {
        try {
          this.fetchSuccess(data); // передаём полезные данные в fetchSuccess, дальше по цепочке пойдёт успешный пустой промис
        } catch (error) {
          this.fetchError(error.message); // если что-то пошло не так - дальше по цепочке пойдёт отвергнутый промис
        }
      })
      .catch((error) => {
        this.fetchError(error.userMessage || error.message);
      });
  };

  editClient = (editedClient) => {
    const editedClients = mModules.editClient(this.state.clients, editedClient);
    this.setState({
      clients: editedClients,
      filteredClients: this.filterList(editedClients),
      formMode: 0,
      clientToEdit: null
    });
  };

  addClient = (client) => {
    const addedClients = mModules.addClient(this.state.clients, client);
    this.setState({
      clients: addedClients,
      filteredClients: this.filterList(addedClients),
      formMode: 0,
      clientToEdit: null
    });
  };

  removeClient = (id) => {
    const clients = mModules.removeClient(this.state.clients, id);
    this.setState({clients: clients, filteredClients: this.filterList(clients), formMode: 0});
  };

  onClientEdit = (id) => {
    this.setState({formMode: 1, clientToEdit: this.state.clients.find((client) => client.id === id)});
  };

  onClientAdd = () => {
    const id = this.state.clients.length ? this.state.clients[this.state.clients.length - 1].id + 1 : 1;
    this.setState({formMode: 2, clientToEdit: {id: id, name: '', surName: '', secondName: '', balance: 0, status: ''}})
  };

  closeForm = (formMode) => {
    this.setState({formMode: formMode});
  };

  filterList = (arr, type = this.state.filterMod) => {
    if (type === 'all') {
      return arr;
    } else {
      return arr.filter((item) => type === 'active' ? item.status === 1 : item.status === 0);
    }
  };

  onFilterClick = (e) => {
    const type = e.target.value;
    this.setState({filterMod: type, filteredClients: this.filterList(this.state.clients, type)})
  };

  render() {

    const clients = this.state.clients.map((client) => {
      return <MobileClient path={this.props.path} key={client.id} client={client}/>
    });

    return (

      <div className='mobile'>
        <div className='mobile__filter'>
          <button className='btn filter-all' value='all' onClick={this.onFilterClick}>Все</button>
          <button className='btn filter-active' value='active' onClick={this.onFilterClick}>Активные</button>
          <button className='btn filter-unavail' value='unavail' onClick={this.onFilterClick}>Заблокированные</button>
        </div>
        <table className='mobile__clients'>
          <thead>
          <tr>
            <td>Фамилия</td>
            <td>Имя</td>
            <td>Отчетсво</td>
            <td>Баланс</td>
            <td>Статус</td>
            <td>Управление</td>
          </tr>
          </thead>
          <tbody>
          {
            (!this.state.isLoaded) ?
              <tr>
                <td>Загрузка данных...</td>
              </tr>
              :
              this.state.clients.length ?
                clients
                :
                <tr>
                  <td>Клиентов нет</td>
                </tr>
          }
          </tbody>
        </table>
        <div className='mobile__add'>
          <button className='btn edit-btn add-client' onClick={this.onClientAdd}>Добавить клиента</button>
        </div>
        <div className="mobile__form">
          {
            (this.state.formMode === 1) &&
            <MobileForm key={this.state.clientToEdit.id} btnText='Редактировать' title='Редактировать данные клиента'
                        client={this.state.clientToEdit} formMode={this.state.formMode}/>
          }
          {
            (this.state.formMode === 2) &&
            <MobileForm btnText='Добавить' title='Добавить нового клиента' client={this.state.clientToEdit}
                        formMode={this.state.formMode}/>
          }
        </div>
      </div>
    );
  }
}

const mapStateToProps = function (state) {
  return {
    clients: state.clients.clients,
    error: state.clients.error,
    isLoaded: state.clients.isLoaded
  };
};

export default connect(mapStateToProps)(Mobile);