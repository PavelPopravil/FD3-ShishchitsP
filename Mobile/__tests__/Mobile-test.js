"use strict";

import React from 'react';
import renderer from 'react-test-renderer';

import Mobile from '../src/components/Mobile/index';

const mobileData = require('../src/data.json');

test('работа компонента Mobile', () => {

  const component = renderer.create(
    <Mobile currCompanyName={mobileData.currCompanyName} clients={mobileData.clients}/>
  );

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();

  const addBtn = component.root.find((el) => el.props.className === 'btn edit-btn add-client');

  addBtn.props.onClick(); // Кнопка добавления клиента
  expect(componentTree).toMatchSnapshot();

});
