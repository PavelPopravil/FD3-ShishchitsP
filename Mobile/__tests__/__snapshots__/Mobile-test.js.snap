// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`работа компонента Mobile 1`] = `
<div
  className="mobile container"
>
  <div
    className="mobile__name"
  >
    Velcom
  </div>
  <div
    className="mobile__btn-wrap"
  >
    <button
      className="btn"
      onClick={[Function]}
      value="Velcom"
    >
      Velcom
    </button>
    <button
      className="btn"
      onClick={[Function]}
      value="Mts"
    >
      Mts
    </button>
  </div>
  <div
    className="mobile__filter"
  >
    <button
      className="btn"
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      className="btn"
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      className="btn"
      onClick={[Function]}
      value="unavail"
    >
      Заблокированные
    </button>
  </div>
  <table
    className="mobile__clients"
  >
    <thead>
      <tr>
        <td>
          Фамилия
        </td>
        <td>
          Имя
        </td>
        <td>
          Отчетсво
        </td>
        <td>
          Баланс
        </td>
        <td>
          Статус
        </td>
        <td>
          Управление
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Виталий
        </td>
        <td>
          Новиков
        </td>
        <td>
          Петрович
        </td>
        <td>
          300
        </td>
        <td
          className="avail"
        >
          Активен
        </td>
        <td>
          <button
            className="btn edit-btn"
            onClick={[Function]}
          >
            Редактировать
          </button>
          <button
            className="btn delete-btn"
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Александ
        </td>
        <td>
          Глеб
        </td>
        <td>
          Юрьевич
        </td>
        <td>
          180
        </td>
        <td
          className="avail"
        >
          Активен
        </td>
        <td>
          <button
            className="btn edit-btn"
            onClick={[Function]}
          >
            Редактировать
          </button>
          <button
            className="btn delete-btn"
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Вячеслав
        </td>
        <td>
          Глеб
        </td>
        <td>
          Юрьевич
        </td>
        <td>
          -100
        </td>
        <td
          className="unavail"
        >
          Неактивен
        </td>
        <td>
          <button
            className="btn edit-btn"
            onClick={[Function]}
          >
            Редактировать
          </button>
          <button
            className="btn delete-btn"
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="mobile__add"
  >
    <button
      className="btn edit-btn add-client"
      onClick={[Function]}
    >
      Добавить клиента
    </button>
  </div>
  <div
    className="mobile__form"
  />
</div>
`;

exports[`работа компонента Mobile 2`] = `
<div
  className="mobile container"
>
  <div
    className="mobile__name"
  >
    Velcom
  </div>
  <div
    className="mobile__btn-wrap"
  >
    <button
      className="btn"
      onClick={[Function]}
      value="Velcom"
    >
      Velcom
    </button>
    <button
      className="btn"
      onClick={[Function]}
      value="Mts"
    >
      Mts
    </button>
  </div>
  <div
    className="mobile__filter"
  >
    <button
      className="btn"
      onClick={[Function]}
      value="all"
    >
      Все
    </button>
    <button
      className="btn"
      onClick={[Function]}
      value="active"
    >
      Активные
    </button>
    <button
      className="btn"
      onClick={[Function]}
      value="unavail"
    >
      Заблокированные
    </button>
  </div>
  <table
    className="mobile__clients"
  >
    <thead>
      <tr>
        <td>
          Фамилия
        </td>
        <td>
          Имя
        </td>
        <td>
          Отчетсво
        </td>
        <td>
          Баланс
        </td>
        <td>
          Статус
        </td>
        <td>
          Управление
        </td>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          Виталий
        </td>
        <td>
          Новиков
        </td>
        <td>
          Петрович
        </td>
        <td>
          300
        </td>
        <td
          className="avail"
        >
          Активен
        </td>
        <td>
          <button
            className="btn edit-btn"
            onClick={[Function]}
          >
            Редактировать
          </button>
          <button
            className="btn delete-btn"
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Александ
        </td>
        <td>
          Глеб
        </td>
        <td>
          Юрьевич
        </td>
        <td>
          180
        </td>
        <td
          className="avail"
        >
          Активен
        </td>
        <td>
          <button
            className="btn edit-btn"
            onClick={[Function]}
          >
            Редактировать
          </button>
          <button
            className="btn delete-btn"
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
      <tr>
        <td>
          Вячеслав
        </td>
        <td>
          Глеб
        </td>
        <td>
          Юрьевич
        </td>
        <td>
          -100
        </td>
        <td
          className="unavail"
        >
          Неактивен
        </td>
        <td>
          <button
            className="btn edit-btn"
            onClick={[Function]}
          >
            Редактировать
          </button>
          <button
            className="btn delete-btn"
            onClick={[Function]}
          >
            Удалить
          </button>
        </td>
      </tr>
    </tbody>
  </table>
  <div
    className="mobile__add"
  >
    <button
      className="btn edit-btn add-client"
      onClick={[Function]}
    >
      Добавить клиента
    </button>
  </div>
  <div
    className="mobile__form"
  />
</div>
`;
